File: frontend/src/lib/compare.ts
/**
 * Comparing two runs without inventing a result.
 *
 * The rule everything here obeys: **Elo is only meaningful inside the
 * tournament that produced it.** A 1300 from a run of six hypotheses and a 1300
 * from a run of fifty are not the same number, so the previous UI's merged
 * cross-run leaderboard was not a ranking of anything. There is no merged
 * ranking in this module and none on the page — each run's list is normalised
 * against its own best and worst, and the two sit side by side for a reader to
 * compare, which is a judgement a person can make and arithmetic cannot.
 *
 * What *does* survive the crossing is a directional delta (more active
 * hypotheses, fewer calls) and the Elo **spread** — how far a run separated its
 * best idea from its median one is a property of that run's own tournament.
 */

import type {
  CompareAnalytics,
  CompareDelta,
  CompareGraft,
  DirectionHint,
  HypothesisRow,
} from "../api/types";
import type { Tone } from "./status";

/* --- Deltas ---------------------------------------------------------------
   `direction_hint` used to be read defensively, because this file and the API
   disagreed about how to spell it. They cannot disagree any more: the types
   below are generated from the API's own schema, so the vocabulary here IS the
   server's. What is left is the reading of it.
   ------------------------------------------------------------------------- */

export type DeltaVerdict = "better" | "worse" | "changed" | "same";

export type DeltaView = {
  metric: string;
  base: number;
  challenger: number;
  /** Challenger minus baseline. */
  change: number;
  direction: DirectionHint;
  verdict: DeltaVerdict;
  /** A word, because an arrow and a colour are not enough on their own. */
  label: string;
  arrow: "▲" | "▼" | "–";
  tone: Tone;
};

export function readDelta(delta: CompareDelta): DeltaView {
  const change = delta.challenger - delta.base;
  const direction = delta.direction_hint;

  let verdict: DeltaVerdict = "same";
  if (change !== 0) {
    if (direction === "flat") verdict = "changed";
    else if (direction === "up") verdict = change > 0 ? "better" : "worse";
    else verdict = change < 0 ? "better" : "worse";
  }

  return {
    metric: delta.metric,
    base: delta.base,
    challenger: delta.challenger,
    change,
    direction,
    verdict,
    label: VERDICT_LABEL[verdict],
    arrow: change === 0 ? "–" : change > 0 ? "▲" : "▼",
    tone: VERDICT_TONE[verdict],
  };
}

const VERDICT_LABEL: Record<DeltaVerdict, string> = {
  better: "better",
  worse: "worse",
  changed: "different",
  same: "no change",
};

const VERDICT_TONE: Record<DeltaVerdict, Tone> = {
  better: "go",
  worse: "danger",
  changed: "info",
  same: "neutral",
};

export function readDeltas(deltas: readonly CompareDelta[]): DeltaView[] {
  return deltas.map(readDelta);
}

/** `+3`, `−1.4`, `0`. A signed number is a sentence's worth of meaning here. */
export function formatChange(change: number): string {
  if (change === 0) return "0";
  const rounded = Math.round(change * 10) / 10;
  return `${rounded > 0 ? "+" : "−"}${Math.abs(rounded)}`;
}

export function formatMetricValue(value: number): string {
  return Number.isInteger(value) ? String(value) : String(Math.round(value * 10) / 10);
}

/* --- The headline --------------------------------------------------------- */

export type Verdict = {
  headline: string;
  detail: string;
  tone: Tone;
};

/**
 * One sentence for someone who opened this page to find out whether the change
 * they made helped. It counts only the metrics that carry a direction — rounds
 * and matches judged describe a run, they do not grade it — and it says "no
 * clear winner" rather than manufacturing one.
 */
export function headlineVerdict(
  views: readonly DeltaView[],
  options: { sharedPrompt: boolean },
): Verdict {
  const scored = views.filter(
    (view) => view.verdict === "better" || view.verdict === "worse",
  );
  const better = scored.filter((view) => view.verdict === "better");
  const worse = scored.filter((view) => view.verdict === "worse");

  const caveat = options.sharedPrompt
    ? "Both runs started from the same prompt, so the difference is down to the settings."
    : "These runs did not start from the same prompt, so any difference may be the question rather than the settings.";

  if (scored.length === 0) {
    return {
      headline: "Nothing separates these two runs",
      detail: `No measure with a better-or-worse direction moved. ${caveat}`,
      tone: "neutral",
    };
  }
  if (better.length > worse.length) {
    return {
      headline: `The challenger came out ahead on ${better.length} of ${scored.length} measures`,
      detail: `Ahead on ${list(better)}. ${worse.length > 0 ? `Behind on ${list(worse)}. ` : ""}${caveat}`,
      tone: "go",
    };
  }
  if (worse.length > better.length) {
    return {
      headline: `The baseline came out ahead on ${worse.length} of ${scored.length} measures`,
      detail: `The challenger is behind on ${list(worse)}. ${better.length > 0 ? `Ahead on ${list(better)}. ` : ""}${caveat}`,
      tone: "caution",
    };
  }
  return {
    headline: "No clear winner",
    detail: `The challenger is ahead on ${list(better)} and behind on ${list(worse)}. ${caveat}`,
    tone: "info",
  };
}

function list(views: readonly DeltaView[]): string {
  const names = views.map((view) => view.metric.toLowerCase());
  if (names.length <= 1) return names[0] ?? "nothing";
  return `${names.slice(0, -1).join(", ")} and ${names[names.length - 1]}`;
}

/* --- Ranked lists, normalised inside one run ------------------------------ */

export type RankedRow = {
  hid: string;
  title: string;
  elo: number | null;
  status: string;
  matches: number;
  wins: number;
  rank: number;
  /**
   * Bar length, 0–1, of this hypothesis between its own run's weakest and
   * strongest. It is meaningless across runs, which is exactly why it never
   * leaves the list it was computed for.
   */
  share: number;
};

export type RankedList = {
  rows: RankedRow[];
  /** Which population was ranked: the run's active ideas, or everything it has. */
  basis: "active" | "all" | "none";
  best: number | null;
  worst: number | null;
  total: number;
};

/** The shortest bar still has to be visible, or the last row reads as absent. */
const MIN_SHARE = 0.08;

/**
 * Ranks a run's hypotheses against themselves.
 *
 * An all-rejected run — there is one in the archive, 18 of 18 — has no active
 * hypotheses and no tournament. It falls back to every hypothesis it has and
 * reports `basis: "all"`, so the page can say what it is showing instead of
 * rendering an empty column that looks like a failure.
 */
export function rankWithinRun(rows: readonly HypothesisRow[], limit = 10): RankedList {
  const active = rows.filter((row) => row.status === "active");
  const basis: RankedList["basis"] =
    active.length > 0 ? "active" : rows.length > 0 ? "all" : "none";
  const population = basis === "active" ? active : rows;

  const ranked = [...population].sort((a, b) => (b.elo ?? 0) - (a.elo ?? 0));
  const scores = ranked
    .map((row) => row.elo)
    .filter((elo): elo is number => typeof elo === "number");
  const best = scores.length > 0 ? Math.max(...scores) : null;
  const worst = scores.length > 0 ? Math.min(...scores) : null;
  const span = best !== null && worst !== null ? best - worst : 0;

  return {
    basis,
    best,
    worst,
    total: population.length,
    rows: ranked.slice(0, limit).map((row, index) => ({
      hid: row.hid,
      title: row.title,
      elo: row.elo,
      status: row.status,
      matches: row.matches,
      wins: row.wins,
      rank: index + 1,
      share:
        row.elo === null || best === null || worst === null
          ? MIN_SHARE
          : span === 0
            ? 1
            : MIN_SHARE + (1 - MIN_SHARE) * ((row.elo - worst) / span),
    })),
  };
}

/* --- Graft summaries ------------------------------------------------------
   Null means "this run's engine had no such concept" — a v1 run predates the
   diversity injection entirely. Rendering zeros there would claim it never
   fired, which is a different and false statement.
   ------------------------------------------------------------------------- */

export type GraftView = {
  enabled: boolean;
  firedCount: number;
  collapseEvents: number;
};

export function readGraft(summary: CompareGraft | null | undefined): GraftView | null {
  if (!summary) return null;
  return {
    enabled: summary.enabled,
    firedCount: summary.fired_count,
    collapseEvents: summary.collapse_events,
  };
}

export function hasGraftData(analytics: CompareAnalytics): boolean {
  return (
    readGraft(analytics.graft_summary.baseline) !== null ||
    readGraft(analytics.graft_summary.challenger) !== null
  );
}

/* --- Movement -------------------------------------------------------------
   Two independent runs share no hypothesis ids — `h003` is a per-run counter —
   so movement is matched on title and labelled as such wherever it is shown.
   ------------------------------------------------------------------------- */

export type MovementRow = {
  hid: string;
  title: string;
  rank: number;
  previousRank: number;
  delta: number | null;
};

/** Only a hypothesis the baseline also had can have moved; the rest are new. */
export function readMovement(analytics: CompareAnalytics): MovementRow[] {
  return analytics.movement
    .filter((row) => row.previous_rank !== null)
    .map((row) => ({
      hid: row.hid,
      title: row.title,
      rank: row.rank,
      previousRank: row.previous_rank as number,
      delta: row.delta,
    }));
}

File: backend/app/services/compare.py
"""Comparison analytics between two runs, computed from the database.

Ported from the artifact-walking `api/comparisons.py`, with the statistically invalid part
removed. That endpoint merged both runs' hypotheses into one Elo-ranked leaderboard — but
Elo is only meaningful inside the tournament that produced it, so a 1300 from a run of six
hypotheses and a 1300 from a run of fifty are not the same number and ranking them against
each other invents a result. Comparison here is directional deltas plus title-matched
movement; the two ranked lists the compare page shows come from each run's own
`GET /runs/{id}/hypotheses`, normalized within its own run.

The saved comparison groups are gone with it. A comparison is two run ids, which is what a
URL already is; the tables stay until Revision B drops them.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any
from uuid import UUID

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.db.engine_models import GraftEvent, Match, Run
from app.services.runs import reads

__all__ = ["SameRunComparison", "compare_runs"]


class SameRunComparison(ValueError):
    """A run compared against itself. Every delta would be zero and mean nothing."""


def compare_runs(session: Session, baseline: Run, challenger: Run) -> dict[str, Any]:
    """Build `CompareAnalytics` for two runs."""
    if baseline.id == challenger.id:
        raise SameRunComparison(str(baseline.id))

    base_rows = reads.hypothesis_rows(session, baseline.id)
    challenger_rows = reads.hypothesis_rows(session, challenger.id)

    return {
        "baseline": reads.run_summary(session, baseline),
        "challenger": reads.run_summary(session, challenger),
        # Imported runs have no recorded prompt hash, so this is False for all of them and
        # the page warns rather than implying the two runs answered the same question.
        "shared_prompt": bool(
            baseline.base_prompt_hash
            and challenger.base_prompt_hash
            and baseline.base_prompt_hash == challenger.base_prompt_hash
        ),
        "deltas": _deltas(session, baseline, challenger, base_rows, challenger_rows),
        "movement": _movement(base_rows, challenger_rows),
        "graft_summary": {
            "baseline": _graft(session, baseline),
            "challenger": _graft(session, challenger),
        },
    }


def _deltas(
    session: Session,
    baseline: Run,
    challenger: Run,
    base_rows: Sequence[Mapping[str, Any]],
    challenger_rows: Sequence[Mapping[str, Any]],
) -> list[dict[str, Any]]:
    """The metrics that survive being compared across two independent tournaments.

    Absolute Elo is not among them. `Elo spread` is: the distance from a run's best
    hypothesis to its median one is a property of that run's own tournament, so comparing
    two spreads says something real about how sharply each run separated its ideas.
    """
    base_matches = _completed_matches(session, baseline.id)
    challenger_matches = _completed_matches(session, challenger.id)
    return [
        _delta("Active hypotheses", _count(base_rows, "active"),
               _count(challenger_rows, "active"), "up"),
        _delta("Rejected hypotheses", _count(base_rows, "rejected"),
               _count(challenger_rows, "rejected"), "flat"),
        _delta("Matches judged", base_matches, challenger_matches, "flat"),
        _delta("Rounds", baseline.round, challenger.round, "flat"),
        _delta("Model calls", baseline.calls_used, challenger.calls_used, "down"),
        _delta("Distinct clusters", _clusters(base_rows), _clusters(challenger_rows), "flat"),
        _delta("Elo spread", _elo_spread(base_rows), _elo_spread(challenger_rows), "up"),
    ]


def _delta(metric: str, base: float, challenger: float, direction_hint: str) -> dict[str, Any]:
    return {
        "metric": metric,
        "base": float(base),
        "challenger": float(challenger),
        "direction_hint": direction_hint,
    }


def _movement(
    base_rows: Sequence[Mapping[str, Any]], challenger_rows: Sequence[Mapping[str, Any]]
) -> list[dict[str, Any]]:
    """Where the challenger's leading ideas sat in the baseline, matched by title.

    Titles are the only identity two independent runs share — `hid`s are per-run counters,
    so h003 in one run has nothing to do with h003 in another.
    """
    previous = {
        _key(row["title"]): index + 1
        for index, row in enumerate(base_rows)
        if row["title"].strip()
    }
    movement = []
    for index, row in enumerate(challenger_rows[:10]):
        rank = index + 1
        previous_rank = previous.get(_key(row["title"]))
        movement.append(
            {
                "hid": row["hid"],
                "title": row["title"],
                "rank": rank,
                "previous_rank": previous_rank,
                "delta": None if previous_rank is None else previous_rank - rank,
            }
        )
    return movement


def _graft(session: Session, run: Run) -> dict[str, Any] | None:
    """Null for runs where the concept does not apply, so the UI can hide it entirely.

    Rendering zeros for a v1 run would claim its diversity injection never fired, when in
    fact that engine had no such thing.
    """
    if run.source_version != "v2":
        return None
    collapse_events = session.execute(
        select(func.count()).select_from(GraftEvent).where(GraftEvent.run_id == run.id)
    ).scalar_one()
    graft_state = dict(run.graft_state or {})
    return {
        "enabled": bool((dict(run.config or {}).get("graft") or {}).get("enabled", False)),
        "fired_count": int(graft_state.get("fired_count", 0) or 0),
        "collapse_events": int(collapse_events),
    }


def _completed_matches(session: Session, run_id: UUID) -> int:
    return session.execute(
        select(func.count())
        .select_from(Match)
        .where(Match.run_id == run_id, Match.status == "completed")
    ).scalar_one()


def _count(rows: Sequence[Mapping[str, Any]], status: str) -> int:
    return sum(1 for row in rows if row["status"] == status)


def _clusters(rows: Sequence[Mapping[str, Any]]) -> int:
    return len({row["cluster"] for row in rows if row.get("cluster")})


def _elo_spread(rows: Sequence[Mapping[str, Any]]) -> float:
    active = sorted(
        (row["elo"] for row in rows if row["status"] == "active"), reverse=True
    )
    if len(active) < 2:
        return 0.0
    return round(active[0] - active[len(active) // 2], 1)


def _key(title: str) -> str:
    return " ".join(title.split()).lower()
